### PR TITLE
range proof 

### PR DESCRIPTION
**Pull request by omershlo**
Paillier range proof

### CODE REVIEW

File: benches/proof.rs [hunk 1]
```diff
@@ -32,7 +32,7 @@ pub fn bench_zk_proof_prove_and_verify_1024(b: &mut Bencher) {
 
     b.iter(|| {
         let proof_results = Paillier::prove(&dk, &challenge);
-        let _result = Paillier::verify(&proof_results.unwrap(), &verification_aid);
+        let _result = Paillier::verify_correctKey(&proof_results.unwrap(), &verification_aid);
```
Comment (mortendahl) [inline benches/proof.rs]: with https://github.com/mortendahl/rust-paillier/pull/25 you shouldn't need to do this anymore! you'll simply have to explicitly choose the proof needed, ie `use paillier::proof::CorrectKey;`, and it'll know which `verify` you're thinking about

Comment (omershlo) [reply benches/proof.rs]: Thanks, I will merge #25. I was wandering how come it compiles on @gbenattar machine without renaming 'verify'

Comment (mortendahl) [reply benches/proof.rs]: 👍

File: src/proof/correct_key.rs [hunk 2]
```diff
@@ -93,7 +93,6 @@ impl ProveCorrectKey<EncryptionKey, DecryptionKey> for Paillier
         let y: Vec<_> = (0..STATISTICAL_ERROR_FACTOR)
             .map(|_| BigInt::sample_below(&ek.n))
             .collect();
-
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: please remove this change.

File: src/proof/range_proof.rs [hunk 3]
```diff
@@ -345,4 +346,29 @@ mod tests {
         assert!(result.is_ok());
     }
 
+    #[bench]
+    fn bench_range_proof(b: &mut Bencher){
+        // TODO: bench range for 256bit range.
+        b.iter(|| {
+            /// common:
+            let range = BigInt::from(0xFFFFFFFFFFFFFi64);
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Where is this big int from? In other words, why that value?

File: src/proof/range_proof.rs [hunk 4]
```diff
@@ -345,4 +346,29 @@ mod tests {
...
+            /// prover:
+            let (ek, dk) = test_keypair().keys();
+            /// verifier:
+            let (com,r,e) = Paillier::verifier_commit();
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: (picky) add space -> let (com, r, e)

File: src/proof/range_proof.rs [hunk 5]
```diff
@@ -345,4 +346,29 @@ mod tests {
...
+            /// prover:
+            let (encrypted_pairs, data_and_randmoness_pairs) = Paillier::generate_encrypted_pairs(&ek, &range);
+            /// prover:
+            let secret_r = BigInt::sample_below(&ek.n);
+            let secret_x = BigInt::from(0xFFFFFFFi64);
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Same as above

File: src/proof/range_proof.rs [hunk 6]
```diff
@@ -345,4 +346,29 @@ mod tests {
...
+            /// common:
+            let cipher_x = Paillier::encrypt_with_chosen_randomness(&ek, &RawPlaintext::from( secret_x.clone()), &Randomness(secret_r.clone()));
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Extra space -> [..] from(secret_x.clone()

File: src/proof/range_proof.rs [hunk 7]
```diff
@@ -345,4 +346,29 @@ mod tests {
...
+            // verifer decommits (tested in test_commit_decommit)
+            /// prover:
+            let z_vector= Paillier::generate_proof(&ek,&secret_x,&secret_r,&e,&range,&data_and_randmoness_pairs);
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add spaces: a, b, c, d

File: src/proof/range_proof.rs [hunk 8]
```diff
@@ -345,4 +346,29 @@ mod tests {
...
+            /// verifier:
+            let result = Paillier::verifier_output(&ek,&e,&encrypted_pairs,&z_vector,&range,&cipher_x );
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: same as above

File: src/proof/range_proof.rs (hunk 7 above)
Comment (gbenattar) [reply src/proof/range_proof.rs]: also:
```
let z_vector = 
```

File: src/proof/correct_key.rs [hunk 9]
```diff
@@ -17,6 +17,7 @@ const STATISTICAL_ERROR_FACTOR: usize = 40;
 
 
 #[derive(Debug)]
+// TODO: generelize the error string and move the struct to a common location where all other proofs can use it as well
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: Type: Generalize

File: src/proof/correct_key.rs [hunk 10]
```diff
@@ -62,11 +63,11 @@ pub trait ProveCorrectKey<EK, DK> {
     fn prove(dk: &DK, challenge: &Challenge) -> Result<CorrectKeyProof, ProofError>;
 
     /// Verify proof.
-    fn verify(proof: &CorrectKeyProof, aid: &VerificationAid) -> Result<(), ProofError>;
+    fn verify_correctKey(proof: &CorrectKeyProof, aid: &VerificationAid) -> Result<(), ProofError>;
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: I think you should keep it "snake" case: `verify_correct_key`

File: src/proof/correct_key.rs [hunk 11]
```diff
@@ -62,11 +63,11 @@ pub trait ProveCorrectKey<EK, DK> {
...
 }
 
-fn compute_digest<IT>(values: IT) -> BigInt
-where  IT: Iterator, IT::Item: Borrow<BigInt>
+pub fn compute_digest<IT>(values: IT) -> BigInt
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: Can we extract this in a utility (I know we have this as a task, can we at least add a comment) or a TODO

File: src/proof/correct_key.rs [hunk 12]
```diff
@@ -130,6 +130,8 @@ impl ProveCorrectKey<EncryptionKey, DecryptionKey> for Paillier
         // check x co-prime with n
         if challenge.x.par_iter().any(|xi| BigInt::egcd(&dk.n, xi).0 != BigInt::one()) {
             return Err(ProofError)
+            //TODO: could lead to timing analysis. please follow the poc code: return from the function only once,
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: `// TODO:`

File: src/proof/correct_key.rs [hunk 13]
```diff
@@ -170,7 +172,7 @@ impl ProveCorrectKey<EncryptionKey, DecryptionKey> for Paillier
         Ok(CorrectKeyProof { y_digest })
     }
 
-    fn verify(proof: &CorrectKeyProof, va: &VerificationAid) -> Result<(), ProofError> {
+    fn verify_correctKey(proof: &CorrectKeyProof, va: &VerificationAid) -> Result<(), ProofError> {
```
Comment (gbenattar) [inline src/proof/correct_key.rs]: Same comment for the snake case

File: src/proof/range_proof.rs [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+use ::arithimpl::traits::*;
+use ::BigInteger as BigInt;
+use ::Paillier as Paillier;
+use ::core::*;
+use proof::correct_key::*;
+use ::traits::*;
+use ::{EncryptionKey, DecryptionKey};
+
+use proof::correct_key::ProofError;
+const STATISTICAL_ERROR_FACTOR: usize = 40;
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+const STATISTICAL_ERROR_FACTOR_IN_BYTES: usize = 5;
+const SECURITY_PARAM : usize  = 256;
+const BITS_IN_BYTE: usize = 8;
+pub struct encrypted_pairs {
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+use ::Paillier as Paillier;
+use ::core::*;
+use proof::correct_key::*;
+use ::traits::*;
+use ::{EncryptionKey, DecryptionKey};
+
+use proof::correct_key::ProofError;
+const STATISTICAL_ERROR_FACTOR: usize = 40;
+const STATISTICAL_ERROR_FACTOR_IN_BYTES: usize = 5;
+const SECURITY_PARAM : usize  = 256;
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: extra space. Also magic number here. Can you add a comment about why it is 256?

File: src/proof/range_proof.rs [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+use ::BigInteger as BigInt;
+use ::Paillier as Paillier;
+use ::core::*;
+use proof::correct_key::*;
+use ::traits::*;
+use ::{EncryptionKey, DecryptionKey};
+
+use proof::correct_key::ProofError;
+const STATISTICAL_ERROR_FACTOR: usize = 40;
+const STATISTICAL_ERROR_FACTOR_IN_BYTES: usize = 5;
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Magic number here, can you add a comment why 5?

File: src/proof/range_proof.rs [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+const SECURITY_PARAM : usize  = 256;
+const BITS_IN_BYTE: usize = 8;
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Magic number here, can you add a comment why 8?

File: src/proof/range_proof.rs [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+    r1: Vec<BigInt>,
+    r2: Vec<BigInt>,
+}
+// z_i = (j, x+wi, r*ri modN)
+pub struct masked_witness{
+    j: Vec<u8>,
+    masked_x: Vec<BigInt>,
+    masked_r: Vec<BigInt>,
+}
+pub struct z_vector{
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+    map: Vec<bool>,
+    z1: data_and_randmoness_pairs,
+    z2: masked_witness,
+}
+/// zero-knowledge range proof that a value x<q/3 lies in interval [0,q].
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+/// The verifier is given only c = ENC(ek,x).
+/// The prover has input x, dk, r (randomness used for calculating c)
+/// It is assumed that q is known to both.
+
+/// References:
+/// - Appendix A in [Lindell'17](https://eprint.iacr.org/2017/552)
+/// - section 1.2.2 in [Boudot '00](https://www.iacr.org/archive/eurocrypt2000/1807/18070437-new.pdf)
+
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: remove line

File: src/proof/range_proof.rs [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+    let mut digest = Context::new(&SHA256);
+    let bytes_x: Vec<u8> = x.into();
+    digest.update(&bytes_x);
+
+    let bytes_r: Vec<u8> = r.into();
+    digest.update(&bytes_r);
+    BigInt::from(digest.finish().as_ref())
+}
+
+
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: Remove empty line

File: src/proof/range_proof.rs [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+
+impl RangeProof<EncryptionKey, DecryptionKey,RawPlaintext,Randomness,RawCiphertext> for Paillier {
+
+    fn verifier_commit() -> (BigInt, BigInt, Vec<u8>)
+    {
+
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: remove empty line

File: src/proof/range_proof.rs [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+            .map(|_| random::<u8>())
+            .collect();
+        let mut digest = Context::new(&SHA256);
+        digest.update(&e);
+        let m =  BigInt::from(digest.finish().as_ref());
+        let two = BigInt::from(2u32);
+        let r =  BigInt::sample_below(&two.pow(SECURITY_PARAM as u32));
+        let com = get_hash_commitment(&m, &r);
+        (com,r,e)
+
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: remove empty line

File: src/proof/range_proof.rs [hunk 25]
```diff
@@ -0,0 +1,374 @@
...
+        let mut e: Vec<_> = (0..STATISTICAL_ERROR_FACTOR_IN_BYTES)
+            .map(|_| random::<u8>())
+            .collect();
+        let mut digest = Context::new(&SHA256);
+        digest.update(&e);
+        let m =  BigInt::from(digest.finish().as_ref());
+        let two = BigInt::from(2u32);
+        let r =  BigInt::sample_below(&two.pow(SECURITY_PARAM as u32));
+        let com = get_hash_commitment(&m, &r);
+        (com,r,e)
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 26]
```diff
@@ -0,0 +1,374 @@
...
+
+    }
+    fn generate_encrypted_pairs(ek: &EncryptionKey, range: &BigInt) -> (encrypted_pairs, data_and_randmoness_pairs)
+
+    {
+        let mut encrypted_pairs = encrypted_pairs{
+            c1: Vec::new(),
+            c2: Vec::new(),
+        };
+        let mut data_and_randmoness_pairs = data_and_randmoness_pairs{
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: add empty line above

File: src/proof/range_proof.rs [hunk 27]
```diff
@@ -0,0 +1,374 @@
...
+            map: Vec::new(),
+            z1: data_and_randmoness_pairs,
+            z2: masked_witness,
+        };
+
+        for i in 0..STATISTICAL_ERROR_FACTOR_IN_BYTES
+            {
+                let mut ei = &e[i];
+                for j in 0..BITS_IN_BYTE
+                    {
```
Comment (gbenattar) [inline src/proof/range_proof.rs]: indentation issue

File: src/proof/range_proof.rs (hunk 3 above)
Comment (omershlo) [reply src/proof/range_proof.rs]: just a small number I used for testing. I replaced it now with a 256bit number to make a more realistic scenario and benchmarks

File: src/proof/range_proof.rs (hunk 18 above)
Comment (omershlo) [reply src/proof/range_proof.rs]: I cannot explain it better than bits in a byte

**Comment by mortendahl** [comment]
hi @omershlo, great work on the range proof!

didn't managed to finish but have made some suggestions [here](https://github.com/mortendahl/rust-paillier/tree/range_proof) to perhaps make the code a bit more rustean as well as enable parallel computations. i basically forked you branch, merged in the latest dev, made changes, and pushed here. let's see if i can't finish it tomorrow evening.

i've also added you as a collaborator to this repo so we can just work on the same branch here (eg `range_proof`) -- didn't know git well enough to easily make a PR to your fork but let me know if you prefer it that way instead!

**Comment by omershlo** [comment]
@mortendahl you are humble! you made major progress!.
 It will take me tomorrow to catch-up with your changes. As you suggested, we will continue working on it in `range_proof` branch. I am curious to compare benchmarks to my initial version (2.45sec for 2048bit) [by the way I am not sure at all the benchmarks are not faking]

**Comment by mortendahl** [comment]
I've created another PR using the range_proof branch instead -- okay to close this on @omershlo?

**Pull request closed without merging**